### DISCUSSION TITLE
Weird serial output using clean Tasmota firmware on ESP16E

### DISCUSSION THREAD

**Discussion by Gemikro1** [arendst/Tasmota#18225] [Q&A]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
I flashed a fresh built Tasmota image with no modifications on a custom made device equipped with an generic ESP-16e using a serial programmer (erasing all flash contents before). 
The firmware boots and i can see the boot message at 74880 baud:
```
 ets Jan  8 2013,rst cause:2, boot mode:(3,6)

load 0x4010f000, len 3584, room 16 
tail 0
chksum 0xb0
csum 0xb0
v3969889e
~ld
```
followed at 115200 baud by
```
00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F6, Count 14
00:00:00.055 SER: Set to 5N1 115200 bit/s
```
followed by only garbage
```
0L0L0ÌÁµÅØ¤ ôÕ±K{FF}0L0L0ÌÁ7ÁÙ©¥D -Í­Ñ!ÔZ4­Ñ!d%6©¹ &.ÍÁ®†Ñhk/-¥-ö·´9J06-l­,´Ìº-ºì©K{FF}0L0L0ÌÕ6ÝÊÒ ô¹W´Ê {1H!Ë#¹Wb l 4³Š%¦2N6I·¦·M£¥.Ô½R±Ì¡Ñhk/-µ84™–­†¹s.µê0L0L°æÝ K&¹®´‰µê0L0L0ÍÁ1¡¦ +"h%f%…D)+ ¹ ³ëÑ¡á¢dÍ®Í¸w K4
¥0‘’V³Å˜Á.æÍkê±Ó²¦²ÍÁ±­L:(Ñ´j0-¹'4½—•#K/¹.µê±Ó²¦²ÍÉ¹µŒÓ Í ô¹W´ê¹ ’²É:d²KÅ7'3'5Í°µŒÓ ô¹W4È…)•$D½°¥83É#ÔÉ$%M9(¹  VK{FF}±Ó²¦2s4m #Éi+´K{FF}±Ó²¦3s°V +4…+-Œ) ô¹W´ê¹.¹-ý±Ó²¦3s3& +4ÑÖ£¹®´ê¹ ’²É:d²
```
What causes the serial interface be set to `5N1` ?
I tried to decode this garbage at all resonable baudrates, even using the strange indicated 5N1 but no success.
I can connect to the WifiManager and bring the device into my network as on any other of my Tasmota devices but serial output stays garbage.
Even an OTA update to the standard precomiled Tasmota firmware image gives the same results.
Is there something broken in the serial interface ?

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): _____
- [x] Tasmota binary firmware version number used: 12.4.0.1
  - [x] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: Serial Programmer
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
18:18:18.377 CMD: Status0
18:18:18.405 RSL: STATUS0 = {"Status":{"Module":1,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_8BCF39","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0},"StatusPRM":{"Baudrate":115200,"SerialConfig":"5N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:12:08","StartupUTC":"2023-03-19T17:06:10","Sleep":50,"CfgHolder":4617,"BootCount":34,"BCResetTime":"2023-03-19T17:03:22","SaveCount":44,"SaveAddress":"F9000"},"StatusFWR":{"Version":"12.4.0(tasmota)","BuildDateTime":"2023-02-24T15:32:38","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"372/699"},"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["RALF",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00206000","00044000","00000000"]},"StatusMEM":{"ProgramSize":632,"Free":368,"Heap":24,"ProgramFlashSize":1024,"FlashSize":4096,"FlashChipId":"16405E","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","54000020","00000080"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"},"StatusNET":{"Hostname":"tasmotaxxxxxx","IPAddress":"xxxxxxxx","Gateway":"xxxxxxxx","Subnetmask":"255.255.255.0","DNSServer1":"xxxxxxxxx","DNSServer2":"0.0.0.0","Mac":"xxxxxxxxxx","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":4.8},"StatusMQT":{"MqttHost":"HoPi","MqttPort":8883,"MqttClientMask":"DVES_%06X","MqttClient":"xxxxxx","MqttUser":"xxxxx","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4},"StatusTIM":{"UTC":"2023-03-19T17:18:18","Local":"2023-03-19T18:18:18","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"06:55","Sunset":"19:00"},"StatusSNS":{"Time":"2023-03-19T18:18:18"},"StatusSTS":{"Time":"2023-03-19T18:18:18","Uptime":"0T00:12:08","UptimeSec":728,"Heap":23,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"RALF","BSSId":"E0:28:6D:73:75:CE","Channel":13,"Mode":"11n","RSSI":62,"Signal":-69,"LinkCount":1,"Downtime":"0T00:00:03"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Flash Tasmota, boot device and listen on serial ...

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
Readable output at 115200/8N1 on the serial interface ?

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment]
`SER: Set to 5N1 115200 bit/s` looks like you did some configuration around using serial, and `5N1` is not useful for getting expected output. Maybe try with a standard precompiled build, instead of what you mixed yourself.....

**Comment by Gemikro1** [comment]
That is just the issue, i used the pristine downloaded Tasmota 12.4.0.1, just compiled it without any modifications ...
I _did_ want to add some modifications once the pristine image works on the device but did not change anything yet.

**Comment by sfromis** [comment]
Did you try what I suggested, using a precompiled image?

**Comment by Gemikro1** [comment]
Yes, uploaded the standard Tasmota via OTA from the server (via the minimal image, version on the server was 12.4.0 vs my compiled 12.4.0.1), same issue here with the serial garbage (and the 5N1 message...)

**Comment by Jason2866** [comment]
Erase flash and do a clean install with [Tasmota Web Installer ](https://tasmota.github.io/install/)

**Comment by Gemikro1** [comment]
Ok, erased flash, installed Tasmota ESP8266, now the terminal messages are ok. Why is this not working with the downloaded and compiled Tasmota ? I just unpacked the zip, entered the Wifi and MQTT settings in the user_config_override and compiled ?

**Reply by barbudor** [reply]
As sfromis saif, you probably had something wrong in your flash before
If not by issuing a configuration command that you forgot, may be you initially flashed tasmota without erasing flash
Now with a clean flash you don't any garbage settings in flash and all is good

Always erase flash before an initial flashing

**Comment by Gemikro1** [comment]
That is my default setting in VSCode, i always set PIO settings to erase the flash when uploading self compiled images...
I just uploaded the compiled version again and also this seems to work now.
Weird, never had this issue before, thanks for the help.

**Comment by Jason2866** [comment]
> That is my default setting in VSCode, i always set PIO settings to erase the flash when uploading self compiled images...

There is no default setting in Platformio to Upload and do a erase before! 
I knew this 100% because I implemented this missing function in Tasmota Platformio for VSC 

`pio run -t erase_upload`

**Reply by Gemikro1** [reply]
Ok, i thought the setting under the PIO board configuration for the ESP8266 `Erase Flash : All flash contents` would do exactly this ? 
At least before uploading, there also is a terminal message `Flash will be erased from 0x00000000 to 0x00097fff` when this setting is chosen.

**Reply by Gemikro1** [reply]
I now tried both settings `All flash contents` and `Sketch + WiFi Settings` and in both cases all WiFi, MQTT and TLS settings have been erased. As i had these settings at `All flash contents` from the beginning, the serial error must have come from somewhere else which seems to not have been erased by the PIO upload method...

**Reply by Jason2866** [reply]
Thats (erase) not a option that can be added in a Platformio env. It is a Platformio one in the PIO menu. 
Needs to be activated manually! and after finished the upload command can be fired.
But there is no Pio command in standard Platformio to erase and flash with ONE command given
So you can't have a default setting doing always erase and flash automated.

**Reply by Gemikro1** [reply]
Yes, i agree. But in my case i had made this setting in the PIO menu and selected the corresponding flash size, upload speed and flash mode already from the first upload to the module. 

I just made another try with a pristine ESP8266 fresh out of the sealed bag and flashed the compiled firmware together with the `All flash contents` PIO menu settings (using the programmer from []https://github.com/linshuqin329/ESP-Downloader ).
Result: again the serial error with the 5N1 message.

Seems that the flash erase function in PIO does not cover all necessary areas of the flash, this is at least unexpected.

**Reply by sfromis** [reply]
When you see a message like `Flash will be erased from 0x00000000 to 0x00097fff` that means that flash is **not** erased, only the specifically mentioned segment.

**Reply by Gemikro1** [reply]
Another try using your command `pio run -t erase_upload` gains the same result. 
Only the Tasmota Installer seems to erase every necessary flash area.

**Reply by Gemikro1** [reply]
@sfromis: Can this area be set from PIO somehow to cover the entire flash or do i always need an external tool ?

**Reply by Gemikro1** [reply]
Found one option mentioned on a few pages: `pio run --target erase`, this still does not solve the 5N1 problem though, unfortunately there is also no indication which area it erases during execution.

**Reply by Gemikro1** [reply]
Installing esptool and `python -m esptool --port COMxx --chip esp8266 erase_flash` still gives 5N1.
What does the Tasmota Installer different to get rid of this ?

**Comment by Gemikro1** [comment]
After various tests i now have found one hint: 
It seems that there is no wrong entry or remnants in the flash, there are entries _missing_ !

If i first install tasmota.bin from the server via any method (Tasmotizer, Tasmota Web Installer, esptools) the 5N1 error is gone, even after uploading any of my self compiled versions (Tasmota, Tasmota Lite).
If i then completely erase the flash (i verified that the flash is empty!) and then upload my self compiled version the error is back. 

So the reason cannot be the flash, the reason must be in something missing in the compiled sources !

**Comment by Gemikro1** [comment]
I finally found the reason for the 5N1 error: 
This exactly happens when i include the following two lines into the `user_config_override.h` which i require for communication with a secure MQTT server:
```
#ifndef USE_MQTT_TLS 
#define USE_MQTT_TLS 
```
Inserting these two lines in the otherwise unmodified sources and recompiling causes the 5N1 error after upload.
Uploading the default Tasmota image resets the error but then i do not have TLS enabled.
Please give me a hint on how to get _both_, TLS and Serial to work.

I had several previous projects with TLS enabled using Tasmota 10.0.0 but there i never used the serial interface (shelly 2.5 devices, up and running for some time now without any problems). 
I just now checked the serial interface of one of the devices and the same error is happening there, so it is not an issue of the 12.4.0 version.

**Reply by s-hadinger** [reply]
TLS and serial are unrelated. This is weird. 

I will try this evening, maybe we have something wrong happening when initializing settings

**Reply by Gemikro1** [reply]
Seems that something is configuring serial mode wrong when TLS is activated. 
When i issue `Status 0` with TLS enabled i can also see the configured 5N1. 
In the changelog from version 12.2.0.2 i found an entry: `Default serial bridge configuration from 5N1 to 8N1 regression from v10.1.0.3`, can this be related ?

**Reply by s-hadinger** [reply]
I thought it had been fixed. You didn't enable modbus, right?

**Reply by Gemikro1** [reply]
No, only change was this two lines in the `user_config_override.h` and after some annoying hundred of times compiling yesterday during testing also disabled `lib_i2c`, `lib_display`, `lib_audio` and `lib_rf`to speed up compiling... :(

To be complete: and setting `#define MQTT_FINGERPRINT1  "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00"` to autolearn MQTT fingerprint as recommended under []https://tasmota.github.io/docs/TLS/#fingerprint-validation

**Reply by Gemikro1** [reply]
This is my `user_config_override.h`:

```
/*
  user_config_override.h - user configuration overrides my_user_config.h for Tasmota

  Copyright (C) 2021  Theo Arends

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef _USER_CONFIG_OVERRIDE_H_
#define _USER_CONFIG_OVERRIDE_H_

/*****************************************************************************************************\
 * USAGE:
 *   To modify the stock configuration without changing the my_user_config.h file:
 *   (1) copy this file to "user_config_override.h" (It will be ignored by Git)
 *   (2) define your own settings below
 *
 ******************************************************************************************************
 * ATTENTION:
 *   - Changes to SECTION1 PARAMETER defines will only override flash settings if you change define CFG_HOLDER.
 *   - Expect compiler warnings when no ifdef/undef/endif sequence is used.
 *   - You still need to update my_user_config.h for major define USE_MQTT_TLS.
 *   - All parameters can be persistent changed online using commands via MQTT, WebConsole or Serial.
\*****************************************************************************************************/

/*
Examples :

// -- Master parameter control --------------------
#undef  CFG_HOLDER
#define CFG_HOLDER        4617                   // [Reset 1] Change this value to load SECTION1 configuration parameters to flash

// -- Setup your own Wifi settings  ---------------
#undef  STA_SSID1
#define STA_SSID1         "YourSSID"             // [Ssid1] Wifi SSID

#undef  STA_PASS1
#define STA_PASS1         "YourWifiPassword"     // [Password1] Wifi password

// -- Setup your own MQTT settings  ---------------
#undef  MQTT_HOST
#define MQTT_HOST         "your-mqtt-server.com" // [MqttHost]

#undef  MQTT_PORT
#define MQTT_PORT         1883                   // [MqttPort] MQTT port (10123 on CloudMQTT)

#undef  MQTT_USER
#define MQTT_USER         "YourMqttUser"         // [MqttUser] Optional user

#undef  MQTT_PASS
#define MQTT_PASS         "YourMqttPass"         // [MqttPassword] Optional password

// You might even pass some parameters from the command line ----------------------------
// Ie:  export PLATFORMIO_BUILD_FLAGS='-DUSE_CONFIG_OVERRIDE -DMY_IP="192.168.1.99" -DMY_GW="192.168.1.1" -DMY_DNS="192.168.1.1"'

#ifdef MY_IP
#undef  WIFI_IP_ADDRESS
#define WIFI_IP_ADDRESS     MY_IP                // Set to 0.0.0.0 for using DHCP or enter a static IP address
#endif

#ifdef MY_GW
#undef  WIFI_GATEWAY
#define WIFI_GATEWAY        MY_GW                // if not using DHCP set Gateway IP address
#endif

#ifdef MY_DNS
#undef  WIFI_DNS
#define WIFI_DNS            MY_DNS               // If not using DHCP set DNS IP address (might be equal to WIFI_GATEWAY)
#endif

#ifdef MY_DNS2
#undef  WIFI_DNS2
#define WIFI_DNS2           MY_DNS2              // If not using DHCP set DNS IP address (might be equal to WIFI_GATEWAY)
#endif

// !!! Remember that your changes GOES AT THE BOTTOM OF THIS FILE right before the last #endif !!!
*/


#ifndef USE_MQTT_TLS 
#define USE_MQTT_TLS                             // Use TLS for MQTT connection (+34.5k code, +7.0k mem and +4.8k additional during connection handshake)
#endif

#ifdef MQTT_FINGERPRINT1
#undef MQTT_FINGERPRINT1
#define MQTT_FINGERPRINT1  "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00"
#endif

#endif  // _USER_CONFIG_OVERRIDE_H_

```

**Comment by joba-1** [comment]
When you switch standard tasmota images and self compiled: did you always erase flash in between? I guess, self compiled could move/change settings area.

**Reply by arendst** [reply]
If that's the case then something is seriously wrong. Tasmota ALWAYS makes sure the settings area is 4096 bytes.

**Reply by Gemikro1** [reply]
I curretnly use Tasmotizer to flash as you can easily set the WiFi settings (if serial is working after flashing :( ... ) and have checked the `erase before flashing` option which should be reliable. But also erasing flash using PIO commands `Erase Flash and Upload`or erasing using esptool.py directly gain the same results. 

I currently do not think anymore that this is a problem with flash area, it seems to come from compilation and settings that get mixed up. This should be easy to prove by somebody compiling and uploading the default Tasmota to any ESP8266 with `USE_MQTT_TLS` enabled.
I already tried using (several) generic ESP16e, one Wemos D1 Mini and a Shelly 2.5, same result.
Would also try with an ESP32 if i could find out where i have hidden those ...

**Reply by s-hadinger** [reply]
I don't have an ESP8266 at hand for now, I will try later.

**Reply by Gemikro1** [reply]
One more try using `board = esp8266_4M2M` to test a different flash configuration (my module has 4MB) gains same results.

**Reply by s-hadinger** [reply]
I just compiled a version with `#define USE_MQTT_TLS`, erased flash and flashed the firmware with esptool, no problem. It does not set to 5N1.

I can only suppose that one of your flashing method injected wrong parameters, or that you flash is faulty

**Reply by Gemikro1** [reply]
On more than 5 different devices ? I also tried several different flashing methods, via Serial Programmer, ESP Downloader for the direct module programming, Tasmotizer etc.
Did you use the `user_config_override` i posted ? Which exact configuration did you use ?

**Reply by Gemikro1** [reply]
And apart from the serial interface everything is working perfectly, i can connect to MQTT TLS, i receive status messages (which can be seen on the serial monitor but only as garbage).

**Reply by s-hadinger** [reply]
No, I just changed in my_user_config.h. Let me try with your file

**Reply by s-hadinger** [reply]
Aha. Indeed it switches to 5N1 with your file. Interesting...

**Reply by Gemikro1** [reply]
But why ????

**Reply by s-hadinger** [reply]
It's because your fingerprint has a wrong format.

```
#define MQTT_FINGERPRINT1  "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00"
```

It should be in binary, not as a string
```
#define MQTT_FINGERPRINT1      0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00  // [MqttFingerprint1] (auto-learn)
```

**Reply by s-hadinger** [reply]
This was changed 3 years ago. https://github.com/arendst/Tasmota/commit/b525e5c72021d22fa6380a7b519eec178e31ee0d

I'm afraid there might be some obsolete documentation still.

**Reply by Gemikro1** [reply]
Ok, explains a lot. This is a result if you just copy some lines from web pages without watching.
Maybe the doc under `https://tasmota.github.io/docs/TLS/#fingerprint-validation` should be updated corresponding to this...
Could have saved hours :(

**Comment by Gemikro1** [comment]
Did a test now with `#define MQTT_FINGERPRINT1      0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00  // [MqttFingerprint1] (auto-learn)` and as expected it is working.
Thanks a lot for the help, was already so fixed on something else that i would never have found this.
I will close the discussion.

**Reply by s-hadinger** [reply]
Documentation fixed

**Reply by Gemikro1** [reply]
Perfect, thanks a lot !
